In [2]:
import pandas as pd
import time

In [2]:
file_start_time = time.time()

In [3]:
# header = ['_id', 'type', 'mmsi', 'status', 'turn', 'speed', 'accuracy', 'lon', 'lat', 'course', 'heading', 'time', 'imo', 'callsign', 'shipname', 'shiptype', 'to_bow',
#           'to_stern', 'to_port', 'to_starboard', 'month', 'day', 'hour', 'minute', 'draught', 'destination', 'destination_country',	'shiptype_vessel',	
#           'vessel_flag', 'vessel_flag_continent',' vessel_flag_country_code\r']
#         # tbc header name from country_destination

In [6]:
start_time = time.time()

# chunk = pd.read_csv('cleaned_type5.csv',chunksize=100000,lineterminator='\n')
df = pd.read_excel('hourly weather_singapore 2023-09-01 to 2023-09-30.xlsx')
# df = pd.concat(chunk)

# start_time = time.time()
end_time = time.time()
read_time =   end_time - start_time
print("Read time:",read_time)


Read time: 0.5077319145202637


In [29]:
df.tail()

name            datetime  temp  feelslike   dew  humidity  precip  \
715  singapore 2023-09-30 19:00:00  28.0       30.5  21.6     68.27     0.0   
716  singapore 2023-09-30 20:00:00  27.9       31.0  22.7     73.28     0.0   
717  singapore 2023-09-30 21:00:00  27.1       30.1  23.7     82.11     0.0   
718  singapore 2023-09-30 22:00:00  27.0       30.3  24.3     85.20     0.0   
719  singapore 2023-09-30 23:00:00  27.0       30.0  23.8     82.48     0.0   

     precipprob preciptype  snow  ...  sealevelpressure  cloudcover  \
715           0        NaN   0.0  ...            1010.1        88.0   
716           0        NaN   NaN  ...            1010.7        89.2   
717           0        NaN   0.0  ...            1011.1        89.1   
718           0        NaN   0.0  ...            1012.0        89.1   
719           0        NaN   0.0  ...            1012.0        89.2   

     visibility  solarradiation  solarenergy  uvindex  severerisk  \
715         8.6              31          0.1        0          30   
716         9.1               0          0.0        0          10   
717         9.1               0          0.0        0          10   
718         9.1               0          0.0        0          10   
719         8.8               0          0.0        0          10   

           conditions                 icon  \
715  Partially cloudy  partly-cloudy-night   
716  Partially cloudy  partly-cloudy-night   
717  Partially cloudy  partly-cloudy-night   
718  Partially cloudy  partly-cloudy-night   
719  Partially cloudy  partly-cloudy-night   

                                              stations  
715  48698099999,48679099999,WSAP,48694099999,48699...  
716  48698099999,48679099999,WSAP,48694099999,48699...  
717  48698099999,48679099999,WSAP,48694099999,48699...  
718  48698099999,48679099999,WSAP,48694099999,48699...  
719  48698099999,48679099999,WSAP,48694099999,48699...  

[5 rows x 24 columns]

In [8]:
df.describe()

datetime        temp   feelslike         dew    humidity  \
count                  720  720.000000  720.000000  720.000000  720.000000   
mean   2023-09-15 23:30:00   28.638194   32.534583   24.214306   77.861806   
min    2023-09-01 00:00:00   24.000000   24.000000   20.400000   50.650000   
25%    2023-09-08 11:45:00   27.400000   31.000000   23.600000   69.467500   
50%    2023-09-15 23:30:00   28.400000   33.000000   24.300000   79.410000   
75%    2023-09-23 11:15:00   30.000000   35.200000   24.700000   85.695000   
max    2023-09-30 23:00:00   33.000000   39.900000   26.300000   99.300000   
std                    NaN    1.983443    3.710795    0.900211   10.847429   

           precip  precipprob   snow  snowdepth    windgust   windspeed  \
count  720.000000  720.000000  600.0        0.0  720.000000  720.000000   
mean     0.202025    3.888889    0.0        NaN   19.026667   10.817778   
min      0.000000    0.000000    0.0        NaN    1.400000    3.300000   
25%      0.000000    0.000000    0.0        NaN   14.000000    7.300000   
50%      0.000000    0.000000    0.0        NaN   19.800000    9.900000   
75%      0.000000    0.000000    0.0        NaN   24.100000   13.900000   
max     41.673000  100.000000    0.0        NaN   46.400000   32.800000   
std      2.148192   19.346454    0.0        NaN    6.767155    4.477220   

          winddir  sealevelpressure  cloudcover  visibility  solarradiation  \
count  720.000000        720.000000  720.000000  720.000000      720.000000   
mean   138.854167       1009.909861   84.974861    9.569167      260.390278   
min      1.000000       1005.700000   49.600000    2.300000        0.000000   
25%     91.000000       1009.000000   88.000000    9.500000        0.000000   
50%    123.000000       1010.000000   88.550000   10.000000        1.500000   
75%    167.250000       1011.000000   89.100000   10.000000      547.000000   
max    360.000000       1013.700000   99.800000   10.700000     1018.000000   
std     80.034534          1.434989    8.786748    1.088092      346.503134   

       solarenergy     uvindex  severerisk  
count   720.000000  720.000000  720.000000  
mean      0.938056    2.613889   21.387500  
min       0.000000    0.000000    3.000000  
25%       0.000000    0.000000   10.000000  
50%       0.000000    0.000000   30.000000  
75%       2.000000    5.250000   30.000000  
max       3.700000   10.000000   60.000000  
std       1.247026    3.472255   12.874752

In [11]:
df.isna().sum()

name                  0
datetime              0
temp                  0
feelslike             0
dew                   0
humidity              0
precip                0
precipprob            0
preciptype          636
snow                120
snowdepth           720
windgust              0
windspeed             0
winddir               0
sealevelpressure      0
cloudcover            0
visibility            0
solarradiation        0
solarenergy           0
uvindex               0
severerisk            0
conditions            0
icon                  0
stations              0
dtype: int64

In [15]:
df.dtypes

name                        object
datetime            datetime64[ns]
temp                       float64
feelslike                  float64
dew                        float64
humidity                   float64
precip                     float64
precipprob                   int64
preciptype                  object
snow                       float64
snowdepth                  float64
windgust                   float64
windspeed                  float64
winddir                      int64
sealevelpressure           float64
cloudcover                 float64
visibility                 float64
solarradiation               int64
solarenergy                float64
uvindex                      int64
severerisk                   int64
conditions                  object
icon                        object
stations                    object
dtype: object

In [17]:
df.conditions.unique()

array(['Partially cloudy', 'Rain, Partially cloudy', 'Overcast',
       'Rain, Overcast'], dtype=object)

In [18]:
df.icon.unique()

array(['partly-cloudy-night', 'partly-cloudy-day', 'rain', 'cloudy'],
      dtype=object)

In [20]:
df.stations.value_counts()

stations
48698099999,48679099999,WSAP,48694099999,48699099999,WSSL    690
WSAP,WSSL                                                     16
48698099999,48679099999,WSAP,48694099999,48699099999           2
WSAP,48694099999,48699099999,WSSL                              2
WSAP,48699099999,WSSL                                          2
48698099999,WSAP,WSSL                                          2
48698099999,48679099999,WSAP,48699099999,WSSL                  2
48698099999,48679099999,WSAP,WSSL                              1
48679099999,WSAP,WSSL                                          1
48698099999,48679099999,48694099999,48699099999                1
WSAP,48694099999,WSSL                                          1
Name: count, dtype: int64

In [6]:
start_time = time.time()
df.isna().sum()
end_time = time.time()
execution_time =   end_time - start_time
print("Execution time:",execution_time)

Execution time: 0.3024575710296631


In [21]:
df.shape

(720, 24)

In [ ]:
# to extract windspeed, visibility, pressure

# clean weather

In [23]:
df['datetime'] = pd.to_datetime(df['datetime'])

In [25]:
daily_data = df.groupby(pd.Grouper(key='datetime', freq='D')).agg({
    'windspeed': 'mean',
    'visibility': 'mean',
    'sealevelpressure': 'mean'
}).reset_index()

In [30]:
daily_data.tail()

datetime  windspeed  visibility  sealevelpressure
25 2023-09-26  13.537500    9.500000       1009.270833
26 2023-09-27  13.425000   10.041667       1009.141667
27 2023-09-28  12.620833   10.058333       1009.675000
28 2023-09-29   9.695833    9.445833       1011.133333
29 2023-09-30   8.795833    8.133333       1011.687500

In [28]:
daily_data.to_csv('weather_data.csv')

# Cleaning steps
1. Drop message id col
2. Split by msg type (123, 5, 1819)
3. Remove MMSI less than 9 digits
4. Extract SG destination
   
For 123 and 1819:
1. convert status to int

For 5:

# NA
1. Callsign drop non alpha-numeric (callsign ~ '^[A-Za-z0-9]+$'  and > 0)
2. Convert shiptype to int
4. Drop month, day, hour minute


In [8]:
clean_start_time = time.time()

In [9]:
df_clean = df.drop('_id', axis = 1)

In [10]:
df_clean.shape

(1417249, 18)

In [11]:
# clean mmsi shorter than 9 digits
df_clean = df_clean[df_clean.mmsi > 99999999]

In [12]:
df_clean.shape

(1417136, 18)

In [13]:
# start_time = time.time()
# df_type123 = df_clean[df_clean.type <= 3][['mmsi', 'status', 'turn', 'speed', 'accuracy', 'lon', 'lat', 'course', 'heading', 'time']]
# print(df_type123.shape)
# end_time = time.time()
# execution_time =   end_time - start_time
# print("Execution time:",execution_time)

In [14]:
start_time = time.time()
df_type5 = df_clean[df_clean.type == 5][['mmsi', 'time', 'draught', 'destination_country']] # cleaned destination 
print(df_type5.shape)
end_time = time.time()
execution_time =   end_time - start_time
print("Execution time:",execution_time)

(1417136, 4)
Execution time: 0.3599581718444824


In [15]:
# start_time = time.time()
# df_type1819 = df_clean[df_clean.type >= 18][['mmsi', 'speed', 'accuracy', 'lon', 'lat', 'course', 'time']]
# print(df_type1819.shape)
# end_time = time.time()
# execution_time =   end_time - start_time
# print("Execution time:",execution_time)

# Clean df_type 123

In [16]:
# # change status to type int
# df_type123['status'] = df_type123['status'].astype(int)

# Clean df_type 5

# Clean df_type 1819

In [17]:
clean_end_time = time.time()
clean_time =   clean_end_time - clean_start_time
print("Cleaning time:",clean_time)

Cleaning time: 0.7735941410064697


# Destination cleaning (NA)

In [18]:
# # Singapore location cleaning
# keywords = ['sg', 'singapore', 'sin', 'sgp', 'PEBGA', 'PEBGB', 'PSBG', 'PWBGA', 'PWBGB', 'PJSB', 'PGBG', 'ACBTH', 'ACGP', 'AEBA', 'AEBB', 'AEBC', 'AEPB', 'AEPBB', 'AEPBC', 'AESPA', 'ASCA', 'ASCB', 'AMOW', 'AEW', 'AEPA', 'AEHA', 'AEHC', 'AEHB']

# # Function to clean destination data using input
# def clean_destination(destination):
#     destination = str(destination)
#     cleaned_destination = [word.upper() for word in keywords if word.lower() in destination.lower()]
#     if cleaned_destination:
#         return 'SG'
        
# # Apply the function to the 'destination' column and create a new column 'destination_clean'
# df_clean['destination_clean'] = df_clean['destination'].apply(clean_destination)

# # drop destination column
# # df_clean = df_clean.drop('destination', axis = 1)

In [19]:
# df_clean.destination_clean.value_counts()

In [20]:
# countries_keywords = {
#     'SG': ['sg', 'singapore', 'sin', 'sgp'],  # Singapore
#     'US': ['us', 'usa', 'america'],  # United States
#     'CN': ['cn', 'china', 'prc'],  # China
#     'JP': ['jp', 'japan'],  # Japan
#     'KR': ['kr', 'south korea'],  # South Korea
#     'DE': ['de', 'germany'],  # Germany
#     'NL': ['nl', 'netherlands'],  # Netherlands
#     'AE': ['ae', 'united arab emirates', 'uae'],  # United Arab Emirates
#     'HK': ['hk', 'hong kong'],  # Hong Kong
#     'MY': ['my', 'malaysia'],  # Malaysia
#     'BR': ['br', 'brazil'],  # Brazil
#     'AU': ['au', 'australia'],  # Australia
#     'ID': ['id', 'indonesia'],  # Indonesia
#     'TH': ['th', 'thailand'],  # Thailand
#     'VN': ['vn', 'vietnam'],  # Vietnam
#     'PH': ['ph', 'philippines'],  # Philippines
#     'MM': ['mm', 'myanmar'],  # Myanmar
#     'KH': ['kh', 'cambodia'],  # Cambodia
#     'LA': ['la', 'lao', 'laos'],  # Laos
#     'BN': ['bn', 'brunei'],  # Brunei
#     'TL': ['tl', 'timor-leste'],  # Timor-Leste
#     'IN': ['in', 'india'],  # India
#     'PK': ['pk', 'pakistan'],  # Pakistan
#     'BD': ['bd', 'bangladesh'],  # Bangladesh
#     'LK': ['lk', 'sri lanka'],  # Sri Lanka
#     'NP': ['np', 'nepal'],  # Nepal
#     'MM': ['mm', 'myanmar'],  # Myanmar
#     # Add more countries and their keywords as needed
# }

In [21]:
# start_time = time.time()

# for code, keywords in countries_keywords.items():
#     def clean_destination(destination):
#         destination = str(destination)
#         cleaned_destination = [word.upper() for word in keywords if word.lower() in destination.lower()]
#         if cleaned_destination:
#             return code

#     df_clean['destination_clean'] = df_clean['destination'].apply(clean_destination)

# end_time = time.time()
# execution_time =   end_time - start_time
# print("Execution time:",execution_time)
# df_clean.destination_clean.value_counts()

In [22]:
# def clean_destination(row):
#     destination = str(row['destination'].lower())
#     for code, keywords in countries_keywords.items():
#         if any(keyword in destination for keyword in keywords):
#             return code.upper()
#     return None

In [23]:
# # Apply the clean_destination function to create 'destination_clean' column
# df_clean['destination_clean'] = df_clean.apply(clean_destination, axis=1)

In [24]:
# df_clean.destination_clean.value_counts()

# All cleaning steps completed, save to csv

In [25]:
# df_clean.to_csv('df_clean.csv')

In [26]:
# df_type123.to_csv('df_type123.csv')

In [27]:
df_type5.to_csv('df_type5.csv')

In [30]:
df_type5.head()

mmsi                    time  draught destination_country
0  205559000  2023-09-09 18:07:29+08     13.7           SINGAPORE
1  205083000  2023-09-09 15:35:18+08      5.2           SINGAPORE
2  205559000  2023-09-09 09:31:29+08     16.9           SINGAPORE
3  205559000  2023-09-09 06:37:27+08     16.9           SINGAPORE
4  205559000  2023-09-09 08:19:25+08     16.9           SINGAPORE

In [28]:
# df_type1819.to_csv('df_type1819.csv')

In [29]:
file_end_time = time.time()
file_execution_time = file_end_time - file_start_time
print("Read time:",read_time)
print("Cleaning time:",clean_time)
print("File run time:",file_execution_time)

Read time: 2.7133119106292725
Cleaning time: 0.7735941410064697
File run time: 6.264193534851074
